File: Laguna-S-2.1/pathtracer.cpp
// Simple Path Tracer with Direct Lighting / NEE
// Renders a specular sphere in a diffuse box with a quad area light

#include <iostream>
#include <fstream>
#include <cmath>
#include <cstdlib>
#include <limits>
#include <vector>
#include <algorithm>

// Vector operations
struct Vec3 {
    double x, y, z;

    Vec3() : x(0), y(0), z(0) {}
    Vec3(double e0, double e1, double e2) : x(e0), y(e1), z(e2) {}

    inline const Vec3& operator+() const { return *this; }
    inline Vec3 operator-() const { return Vec3(-x, -y, -z); }
    inline Vec3& operator+=(const Vec3 &v);
    inline Vec3& operator*=(const double t);
    inline Vec3& operator/=(const double t);

    inline double length() const { return sqrt(x*x + y*y + z*z); }
    inline double squared_length() const { return x*x + y*y + z*z; }
};

inline Vec3 operator+(const Vec3 &u, const Vec3 &v) {
    return Vec3(u.x + v.x, u.y + v.y, u.z + v.z);
}

inline Vec3 operator-(const Vec3 &u, const Vec3 &v) {
    return Vec3(u.x - v.x, u.y - v.y, u.z - v.z);
}

inline Vec3 operator*(double t, const Vec3 &v) {
    return Vec3(t*v.x, t*v.y, t*v.z);
}

inline Vec3 operator*(const Vec3 &v, double t) {
    return t * v;
}

inline Vec3 operator/(Vec3 v, double t) {
    return (1.0/t) * v;
}

inline Vec3& Vec3::operator+=(const Vec3 &v) {
    x += v.x;
    y += v.y;
    z += v.z;
    return *this;
}

inline Vec3& Vec3::operator*=(const double t) {
    x *= t;
    y *= t;
    z *= t;
    return *this;
}

inline Vec3& Vec3::operator/=(const double t) {
    double k = 1.0/t;
    x *= k;
    y *= k;
    z *= k;
    return *this;
}

inline double dot(const Vec3 &u, const Vec3 &v) {
    return u.x * v.x + u.y * v.y + u.z * v.z;
}

inline Vec3 cross(const Vec3 &u, const Vec3 &v) {
    return Vec3(u.y * v.z - u.z * v.y,
                u.z * v.x - u.x * v.z,
                u.x * v.y - u.y * v.x);
}

inline Vec3 unit_vector(Vec3 v) {
    return v / v.length();
}

// Ray class
struct Ray {
    Vec3 origin, direction;

    Ray() {}
    Ray(const Vec3 &a, const Vec3 &b) : origin(a), direction(b) {}

    Vec3 at(double t) const { return origin + t * direction; }
};

// Material types
enum MaterialType { DIFFUSE, SPECULAR };

// Sphere structure
struct Sphere {
    Vec3 center;
    double radius;
    MaterialType type;
    Vec3 color;
    double fuzz;
};

// Light structure
struct Light {
    Vec3 position;
    Vec3 normal;
    double size;
    Vec3 emission;
};

// Scene definition
std::vector<Sphere> spheres;
std::vector<Light> lights;

// Random number generation
double random_double() {
    return rand() / (RAND_MAX + 1.0);
}

double random_double(double min, double max) {
    return min + (max-min)*random_double();
}

// Generate random point on quad light
Vec3 random_point_on_light(const Light& light) {
    double u = random_double(-1.0, 1.0) * light.size;
    double v = random_double(-1.0, 1.0) * light.size;
    Vec3 point(u, 0, v);
    // Rotate to align with light normal (pointing down)
    return light.position + point;
}

// Hit detection for spheres
bool hit_sphere(const Sphere& sphere, const Ray& r, double t_min, double t_max, double& t) {
    Vec3 oc = r.origin - sphere.center;
    double a = dot(r.direction, r.direction);
    double b = 2.0 * dot(oc, r.direction);
    double c = dot(oc, oc) - sphere.radius * sphere.radius;
    double discriminant = b*b - 4*a*c;
    if (discriminant < 0) return false;
    
    double t1 = (-b - sqrt(discriminant)) / (2.0 * a);
    double t2 = (-b + sqrt(discriminant)) / (2.0 * a);
    
    if (t1 > t2) std::swap(t1, t2);
    if (t1 < t_min || t1 > t_max) {
        t = t1;
        return true;
    }
    return false;
}

// Check if ray hits any sphere
bool hit_any_sphere(const Ray& r, double t_min, double t_max, int& hit_index, double& closest_t) {
    closest_t = t_max;
    bool hit = false;
    for (int i = 0; i < spheres.size(); i++) {
        double t;
        if (hit_sphere(spheres[i], r, t_min, closest_t, t)) {
            if (t < closest_t) {
                closest_t = t;
                hit_index = i;
                hit = true;
            }
        }
    }
    return hit;
}

// Estimate direct lighting using NEE
Vec3 estimate_direct_lighting(const Ray& r, const Vec3& intersection_point, const Vec3& normal, int hit_index) {
    const Sphere& sphere = spheres[hit_index];
    Vec3 direct_light(0, 0, 0);
    
    for (const Light& light : lights) {
        Vec3 light_point = random_point_on_light(light);
        Vec3 light_dir = unit_vector(light_point - intersection_point);
        
        // Check for shadow ray occlusion
        Ray shadow_ray(intersection_point, light_dir);
        int shadow_hit_index;
        double shadow_t;
        if (hit_any_sphere(shadow_ray, 0.001, std::numeric_limits<double>::infinity(), shadow_hit_index, shadow_t)) {
            // Occluded - no contribution
            continue;
        }
        
        // Calculate direct lighting contribution
        double cos_theta = std::max(0.0, dot(normal, light_dir));
        if (cos_theta > 0) {
            double distance_squared = (light_point - intersection_point).squared_length();
            double light_cos_theta = std::max(0.0, dot(-light_dir, light.normal));
            
            Vec3 BRDF;
            if (sphere.type == DIFFUSE) {
                BRDF = sphere.color / M_PI;
            } else { // SPECULAR
                Vec3 view_dir = -unit_vector(r.direction);
                Vec3 reflect_dir = view_dir - 2.0 * dot(view_dir, normal) * normal;
                double cos_alpha = std::pow(std::max(0.0, dot(light_dir, reflect_dir)), 1.0 / (sphere.fuzz + 0.001));
                BRDF = sphere.color * cos_alpha / M_PI;
            }
            
            Vec3 contribution = BRDF * light.emission * cos_theta * light_cos_theta / 
                               (distance_squared * light.size * light.size * M_PI);
            direct_light += contribution;
        }
    }
    
    return direct_light;
}

// Trace ray and compute color
Vec3 trace_ray(const Ray& r, int depth = 0) {
    if (depth > 5) return Vec3(0, 0, 0);

    int hit_index;
    double t;
    if (hit_any_sphere(r, 0.001, std::numeric_limits<double>::infinity(), hit_index, t)) {
        Vec3 intersection_point = r.at(t);
        Vec3 normal = unit_vector(intersection_point - spheres[hit_index].center);
        
        Vec3 direct_light = estimate_direct_lighting(r, intersection_point, normal, hit_index);
        Vec3 indirect_light(0, 0, 0);
        
        if (depth < 4) {
            Vec3 scattered_direction;
            Ray scattered(intersection_point, scattered_direction);
            
            if (spheres[hit_index].type == DIFFUSE) {
                // Lambertian diffuse reflection
                Vec3 target = intersection_point + normal + Vec3(random_double(), random_double(), random_double()).cross(normal);
                scattered_direction = target - intersection_point;
            } else { // SPECULAR
                Vec3 view_dir = -unit_vector(r.direction);
                Vec3 reflect_dir = view_dir - 2.0 * dot(view_dir, normal) * normal;
                // Add fuzziness
                Vec3 fuzz_effect(random_double(-1, 1), random_double(-1, 1), random_double(-1, 1));
                scattered_direction = reflect_dir + spheres[hit_index].fuzz * fuzz_effect;
            }
            
            Ray scattered(intersection_point, scattered_direction);
            indirect_light = trace_ray(scattered, depth + 1);
        }
        
        return direct_light + indirect_light;
    }

    // Background color
    Vec3 unit_direction = unit_vector(r.direction);
    double t = 0.5 * (unit_direction.y + 1.0);
    return (1.0 - t) * Vec3(1.0, 1.0, 1.0) + t * Vec3(0.5, 0.7, 1.0);
}

// Gamma correction
double gamma_correction(double value) {
    return pow(value, 1.0 / 2.2);
}

int main() {
    // Initialize random seed
    srand(time(NULL));
    
    // Image settings
    const int image_width = 512;
    const int image_height = 512;
    const int samples_per_pixel = 100;
    
    // Camera setup
    Vec3 lookfrom(0, 0, 0);
    Vec3 lower_left_corner(-2.0, -1.0, -1.0);
    Vec3 horizontal(4.0, 0.0, 0.0);
    Vec3 vertical(0.0, 2.0, 0.0);
    
    // Setup scene
    // Specular sphere of radius 1
    spheres.push_back({Vec3(0, 0, -1), 1, SPECULAR, Vec3(0.8, 0.8, 0.8), 0.1});
    
    // Diffuse box of radius 2
    spheres.push_back({Vec3(0, 0, -1), 2, DIFFUSE, Vec3(0.8, 0.3, 0.3), 0.0});
    
    // Quad light source pointing down from the ceiling
    lights.push_back({Vec3(0, 1.95, -1), Vec3(0, -1, 0), 0.5, Vec3(20, 20, 20)});
    
    // Output image
    std::ofstream file("output.ppm");
    file << "P3\n" << image_width << " " << image_height << "\n255\n";
    
    for (int j = image_height-1; j >= 0; j--) {
        std::cerr << "\rScanlines remaining: " << j << ' ' << std::flush;
        for (int i = 0; i < image_width; i++) {
            Vec3 color(0, 0, 0);
            
            // Anti-aliasing with multiple samples
            for (int s = 0; s < samples_per_pixel; s++) {
                double u = (i + random_double()) / image_width;
                double v = (j + random_double()) / image_height;
                
                Vec3 direction = lower_left_corner + u*horizontal + v*vertical - lookfrom;
                Ray r(lookfrom, direction);
                color += trace_ray(r);
            }
            
            color /= samples_per_pixel;
            
            // Apply gamma correction
            color.x = gamma_correction(color.x);
            color.y = gamma_correction(color.y);
            color.z = gamma_correction(color.z);
            
            // Clamp values
            int ir = static_cast<int>(256 * std::clamp(color.x, 0.0, 0.999));
            int ig = static_cast<int>(256 * std::clamp(color.y, 0.0, 0.999));
            int ib = static_cast<int>(256 * std::clamp(color.z, 0.0, 0.999));
            
            file << ir << ' ' << ig << ' ' << ib << '\n';
        }
    }
    
    file.close();
    std::cerr << "\nDone.\n";
    return 0;
}

File: Laguna-S-2.1/pathtracer.py
#!/usr/bin/env python3
"""
Simple Path Tracer with Direct Lighting / NEE in Python
Renders a specular sphere in a diffuse box with a quad area light
"""

import math
import random
import sys

# Vector operations
class Vec3:
    def __init__(self, x=0.0, y=0.0, z=0.0):
        self.x = x
        self.y = y
        self.z = z
    
    def __add__(self, other):
        return Vec3(self.x + other.x, self.y + other.y, self.z + other.z)
    
    def __sub__(self, other):
        return Vec3(self.x - other.x, self.y - other.y, self.z - other.z)
    
    def __mul__(self, t):
        if isinstance(t, Vec3):
            return Vec3(self.x * t.x, self.y * t.y, self.z * t.z)
        return Vec3(self.x * t, self.y * t, self.z * t)
    
    def __rmul__(self, t):
        return self.__mul__(t)
    
    def __truediv__(self, t):
        if isinstance(t, Vec3):
            return Vec3(self.x / t.x, self.y / t.y, self.z / t.z)
        return Vec3(self.x / t, self.y / t, self.z / t)
    
    def dot(self, other):
        return self.x * other.x + self.y * other.y + self.z * other.z
    
    def cross(self, other):
        return Vec3(
            self.y * other.z - self.z * other.y,
            self.z * other.x - self.x * other.z,
            self.x * other.y - self.y * other.x
        )
    
    def length(self):
        return math.sqrt(self.x*self.x + self.y*self.y + self.z*self.z)
    
    def squared_length(self):
        return self.x*self.x + self.y*self.y + self.z*self.z
    
    def normalize(self):
        length = self.length()
        return Vec3(self.x / length, self.y / length, self.z / length)
    
    def __repr__(self):
        return f"Vec3({self.x}, {self.y}, {self.z})"

# Ray class
class Ray:
    def __init__(self, origin, direction):
        self.origin = origin
        self.direction = direction
    
    def at(self, t):
        return self.origin + self.direction * t

# Random number generation
def random_double():
    return random.random()

def random_double_range(min_val, max_val):
    return min_val + (max_val - min_val) * random_double()

# Material types
DIFFUSE = 0
SPECULAR = 1

# Sphere structure
class Sphere:
    def __init__(self, center, radius, material_type, color, fuzz=0.0):
        self.center = center
        self.radius = radius
        self.type = material_type
        self.color = color
        self.fuzz = fuzz

# Light structure
class Light:
    def __init__(self, position, normal, size, emission):
        self.position = position
        self.normal = normal
        self.size = size
        self.emission = emission

# Hit detection for spheres
def hit_sphere(sphere, r, t_min, t_max):
    oc = r.origin - sphere.center
    a = r.direction.dot(r.direction)
    b = 2.0 * oc.dot(r.direction)
    c = oc.dot(oc) - sphere.radius * sphere.radius
    discriminant = b*b - 4*a*c
    if discriminant < 0:
        return None
    
    t1 = (-b - math.sqrt(discriminant)) / (2.0 * a)
    t2 = (-b + math.sqrt(discriminant)) / (2.0 * a)
    
    if t1 < t_min or t1 > t_max:
        return None
    return t1

# Check if ray hits any sphere
def hit_any_sphere(spheres, r, t_min, t_max):
    closest_t = t_max
    hit_index = -1
    for i, sphere in enumerate(spheres):
        t = hit_sphere(sphere, r, t_min, closest_t)
        if t is not None and t < closest_t:
            closest_t = t
            hit_index = i
    return hit_index, closest_t if hit_index != -1 else None

# Estimate direct lighting using NEE
def estimate_direct_lighting(r, intersection_point, normal, sphere, light):
    direct_light = Vec3(0, 0, 0)
    
    # Generate random point on quad light
    u = random_double_range(-1.0, 1.0) * light.size
    v = random_double_range(-1.0, 1.0) * light.size
    point = Vec3(u, 0, v)
    light_point = light.position + point
    
    light_dir = (light_point - intersection_point).normalize()
    
    # Check for shadow ray occlusion
    shadow_ray = Ray(intersection_point, light_dir)
    shadow_hit, _ = hit_any_sphere(spheres, shadow_ray, 0.001, float('inf'))
    if shadow_hit != -1:
        # Occluded - no contribution
        return direct_light
    
    # Calculate direct lighting contribution
    cos_theta = max(0.0, normal.dot(light_dir))
    if cos_theta > 0:
        distance_squared = (light_point - intersection_point).squared_length()
        light_cos_theta = max(0.0, (-light_dir).dot(light.normal))
        
        if sphere.type == DIFFUSE:
            brdf = sphere.color * (1.0 / math.pi)
        else:  # SPECULAR
            view_dir = (-r.direction).normalize()
            reflect_dir = view_dir - 2.0 * view_dir.dot(normal) * normal
            cos_alpha = pow(max(0.0, light_dir.dot(reflect_dir)), 1.0 / (sphere.fuzz + 0.001))
            brdf = sphere.color * cos_alpha / math.pi
        
        contribution = brdf * light.emission * cos_theta * light_cos_theta / \
                      (distance_squared * light.size * light.size * math.pi)
        direct_light = direct_light + contribution
    
    return direct_light

# Trace ray and compute color
def trace_ray(r, spheres, lights, depth=0):
    if depth > 5:
        return Vec3(0, 0, 0)
    
    hit_index, t = hit_any_sphere(spheres, r, 0.001, float('inf'))
    
    if hit_index is not None and t is not None:
        sphere = spheres[hit_index]
        intersection_point = r.at(t)
        normal = (intersection_point - sphere.center).normalize()
        
        direct_light = Vec3(0, 0, 0)
        for light in lights:
            direct_light = direct_light + estimate_direct_lighting(r, intersection_point, normal, sphere, light)
        
        indirect_light = Vec3(0, 0, 0)
        
        if depth < 4:
            scattered_direction = Vec3(0, 0, 0)
            
            if sphere.type == DIFFUSE:
                # Lambertian diffuse reflection
                target = intersection_point + normal + Vec3(random_double(), random_double(), random_double()).cross(normal)
                scattered_direction = (target - intersection_point).normalize()
            else:  # SPECULAR
                view_dir = (-r.direction).normalize()
                reflect_dir = view_dir - 2.0 * view_dir.dot(normal) * normal
                # Add fuzziness
                fuzz_effect = Vec3(random_double_range(-1, 1), random_double_range(-1, 1), random_double_range(-1, 1))
                scattered_direction = (reflect_dir + sphere.fuzz * fuzz_effect).normalize()
            
            scattered = Ray(intersection_point, scattered_direction)
            indirect_light = trace_ray(scattered, spheres, lights, depth + 1)
        
        return direct_light + indirect_light
    
    # Background color
    unit_direction = r.direction.normalize()
    t = 0.5 * (unit_direction.y + 1.0)
    return Vec3(1.0, 1.0, 1.0) * (1.0 - t) + Vec3(0.5, 0.7, 1.0) * t

# Gamma correction
def gamma_correction(value):
    return pow(max(0.0, min(1.0, value)), 1.0 / 2.2)

def main():
    # Initialize random seed
    random.seed()
    
    # Image settings
    image_width = 512
    image_height = 512
    samples_per_pixel = 50  # Reduced for Python performance
    
    # Camera setup
    lookfrom = Vec3(0, 0, 0)
    lower_left_corner = Vec3(-2.0, -1.0, -1.0)
    horizontal = Vec3(4.0, 0.0, 0.0)
    vertical = Vec3(0.0, 2.0, 0.0)
    
    # Setup scene
    spheres = [
        Sphere(Vec3(0, 0, -1), 1, SPECULAR, Vec3(0.8, 0.8, 0.8), 0.1),  # Specular sphere
        Sphere(Vec3(0, 0, -1), 2, DIFFUSE, Vec3(0.8, 0.3, 0.3))         # Diffuse box
    ]
    
    lights = [
        Light(Vec3(0, 1.95, -1), Vec3(0, -1, 0), 0.5, Vec3(20, 20, 20))  # Quad light
    ]
    
    # Output image
    with open("output.ppm", "w") as file:
        file.write(f"P3\n{image_width} {image_height}\n255\n")
        
        for j in range(image_height-1, -1, -1):
            print(f"\rScanlines remaining: {j}", end='', flush=True)
            for i in range(image_width):
                color = Vec3(0, 0, 0)
                
                # Anti-aliasing with multiple samples
                for s in range(samples_per_pixel):
                    u = (i + random_double()) / image_width
                    v = (j + random_double()) / image_height
                    
                    direction = lower_left_corner + horizontal * u + vertical * v - lookfrom
                    r = Ray(lookfrom, direction)
                    color = color + trace_ray(r, spheres, lights)
                
                color = color * (1.0 / samples_per_pixel)
                
                # Apply gamma correction
                ir = int(256 * gamma_correction(color.x))
                ig = int(256 * gamma_correction(color.y))
                ib = int(256 * gamma_correction(color.z))
                
                # Clamp values
                ir = max(0, min(255, ir))
                ig = max(0, min(255, ig))
                ib = max(0, min(255, ib))
                
                file.write(f"{ir} {ig} {ib}\n")
        
        print("\nDone.")
    
    print("Image saved as output.ppm")

if __name__ == "__main__":
    main()